### PR TITLE
[#53] State Embedder

### PR DESCRIPTION
**Pull request by JanKapala**
Implementation of the State Embedder.

### CODE REVIEW

File: recommender/engine/rl_agent/preprocessing/state_embedder.py [hunk 1]
```diff
@@ -1,45 +1,136 @@
...
+            self.service_embedder = load_last_module(SERVICES_AUTOENCODER).encoder
 
-        # TODO: implement the rest of initialization (lazy embedders loading)
-        pass
+        if search_phrase_embedder is not None:
+            self.search_phrase_embedder = search_phrase_embedder
+        else:
+            self.search_phrase_embedder = SearchPhraseEmbedder()
+
+        if filters_embedder is not None:
```
Comment (wujuu) [inline recommender/engine/rl_agent/preprocessing/state_embedder.py:61]: I guess you can use this kind of synatx:
`self.search_phrase_embedder = search_phrase_embedder if search_phrase_embedder is not None else  SearchPhraseEmbedder()` if you like.

Comment (wujuu) [reply recommender/engine/rl_agent/preprocessing/state_embedder.py:61]: Here and in other places where module loading can be optional

File: recommender/engine/rl_agent/preprocessing/state_embedder.py [hunk 2]
```diff
@@ -1,45 +1,136 @@
...
-                - search_phrase_tensor of shape [SPE]
+                - user of shape [UE]
+                - services_history of shape [N, SE]
+                - search_phrase of shape [X, SPE]
+                - filters of shape [SE]
+
                 where:
                     - UE is user content tensor embedding dim
-                    - N is user clicked services history length
+                    - N is user clicked services history length - it is not constant
```
Comment (wujuu) [inline recommender/engine/rl_agent/preprocessing/state_embedder.py:96]: maybe it should be lowercase "n" if it's not constant? - same for X

File: recommender/engine/rl_agent/preprocessing/state_embedder.py [hunk 3]
```diff
@@ -1,45 +1,136 @@
...
 
 import torch
+from torch import FloatTensor
+
+from recommender.engine.rl_agent.preprocessing.searchphrase_embedder import (
+    SearchPhraseEmbedder,
+)
+from recommender.engine.pre_agent.models import (
+    load_last_module,
+    USERS_AUTOENCODER,
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engine/rl_agent/preprocessing/state_embedder.py:15]: Agree

File: recommender/engine/rl_agent/preprocessing/state_embedder.py (hunk 1 above)
Comment (JanKapala) [reply recommender/engine/rl_agent/preprocessing/state_embedder.py:61]: or even better: `self.search_phrase_embedder = search_phrase_embedder or SearchPhraseEmbedder()`

File: recommender/engine/rl_agent/preprocessing/state_embedder.py (hunk 2 above)
Comment (JanKapala) [reply recommender/engine/rl_agent/preprocessing/state_embedder.py:96]: If it was a variable/constant in the code - yes, but it is just dimension in the docstring, so it has its own convention.

**Review by wujuu — APPROVED** [review]

**Pull request merged**